hpx/util/pack_traversal.hpp

```diff
@@ -0,0 +1,729 @@
+                    std::forward<T>(container), std::declval<try_traversor>()))
+            {
+                return container_remapping::remap(Strategy{},
+                    std::forward<T>(container), try_traversor{this});
+            }
+
+            /// Match elements which are tuple like and that also may
+            /// satisfy the container requirements
+            /// -> We match tuple like types over container like ones
+            template <bool IsContainer, typename T>
+            auto try_match(
+                container_match_tag<IsContainer, true>, T&& tuple_like)
+                -> decltype(tuple_like_remapping::remap(Strategy{},
+                    std::forward<T>(tuple_like), std::declval<try_traversor>()))
+            {
+                return tuple_like_remapping::remap(Strategy{},
+                    std::forward<T>(tuple_like), try_traversor{this});
+            }
+
+            /// Traverses a single element.
+            ///
+            /// SFINAE helper: Doesn't allow routing through elements,
+            /// that aren't accepted by the mapper
+            template <typename T>
+            auto traverse(Strategy, T&& element)
+                -> decltype(std::declval<mapping_helper>().match(
+                    std::declval<
+                        container_match_of<typename std::decay<T>::type>>(),
+                    std::declval<T>()));
+
+            /// \copybrief traverse
+            template <typename T>
+            auto try_traverse(Strategy, T&& element)
+                -> decltype(std::declval<mapping_helper>().try_match(
+                    std::declval<
+                        container_match_of<typename std::decay<T>::type>>(),
+                    std::declval<T>()))
+            {
+                // We use tag dispatching here, to categorize the type T whether
+                // it satisfies the container or tuple like requirements.
+                // Then we can choose the underlying implementation accordingly.
+                return try_match(
+                    container_match_of<typename std::decay<T>::type>{},
+                    std::forward<T>(element));
+            }
+
+        public:
+            explicit mapping_helper(M mapper)
+              : mapper_(std::move(mapper))
+            {
+            }
+
+            /// \copybrief try_traverse
+            template <typename T>
+            auto init_traverse(Strategy strategy, T&& element)
+                -> decltype(std::declval<mapping_helper>().try_traverse(
+                    strategy, std::declval<T>()))
+            {
+                return try_traverse(strategy, std::forward<T>(element));
+            }
+
+            /// Calls the traversal method for every element in the pack,
+            /// and returns a tuple containing the remapped content.
+            template <typename First, typename Second, typename... T>
+            auto init_traverse(strategy_remap_tag strategy, First&& first,
```

Comment:
GCC has issues with prioritizing the deduction of the function taking only one argument over this one that accepts two or more arguments. This causes endless recursive instantiations.
In order to fix this I spelled both arguments out here.